consensus/hotstuff/eventhandler/event_handler_v2_test.go

```diff
@@ -0,0 +1,527 @@
+		View: es.votingBlock.View + 1, // the vote for the voting blocks will trigger a view change to the next view
+	}
+}
+
+func (es *EventHandlerV2Suite) markInvalidProposal(blockID flow.Identifier) {
+	es.validator.invalidProposals[blockID] = struct{}{}
+}
+
+// a QC for current view triggered view change
+func (es *EventHandlerV2Suite) TestQCBuiltViewChanged() {
+	// voting block exists
+	es.forks.blocks[es.votingBlock.BlockID] = es.votingBlock
+
+	// a qc is built
+	qc := createQC(es.votingBlock)
+
+	// new qc is added to forks
+	// view changed
+	// I'm not the next leader
+	// haven't received block for next view
+	// goes to the new view
+	es.endView++
+	// not the leader of the newview
+	// don't have block for the newview
+	// over
+
+	err := es.eventhandler.OnQCConstructed(qc)
+	require.NoError(es.T(), err, "if a vote can trigger a QC to be built,"+
+		"and the QC triggered a view change, then start new view")
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+}
+
+// a QC for future view triggered view change
+func (es *EventHandlerV2Suite) TestQCBuiltFutureViewChanged() {
+	// voting block exists
+	curView := es.paceMaker.CurView()
+
+	b1 := createBlockWithQC(curView, curView-1)
+	b2 := createBlockWithQC(curView+1, curView)
+	b3 := createBlockWithQC(curView+2, curView+1)
+
+	// a qc is built
+	qc1 := createQC(b1)
+	qc2 := createQC(b2)
+	qc3 := createQC(b3)
+
+	// all three blocks are known
+	es.forks.blocks[b1.BlockID] = b1
+	es.forks.blocks[b2.BlockID] = b2
+	es.forks.blocks[b3.BlockID] = b3
+
+	err := es.eventhandler.OnQCConstructed(qc3)
+	endView := b3.View + 1 // next view
+	require.NoError(es.T(), err, "if a vote can trigger a QC to be built,"+
+		"and the QC triggered a view change, then start new view")
+	require.Equal(es.T(), endView, es.paceMaker.CurView(), "incorrect view change")
+
+	// old QCs won't trigger view change
+	err = es.eventhandler.OnQCConstructed(qc2)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), endView, es.paceMaker.CurView(), "incorrect view change")
+
+	err = es.eventhandler.OnQCConstructed(qc1)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), endView, es.paceMaker.CurView(), "incorrect view change")
+}
+
+// in the newview, I'm not the leader, and I have the cur block,
+// and the block is not a safe node, and I'm the next leader, and no qc built for this block.
+func (es *EventHandlerV2Suite) TestInNewView_NotLeader_HasBlock_NoVote_IsNextLeader_NoQC() {
+	// voting block exists
+	es.forks.blocks[es.votingBlock.BlockID] = es.votingBlock
+	// a qc is built
+	qc := createQC(es.votingBlock)
+	// viewchanged
+	es.endView++
+	// not leader for newview
+
+	// has block for newview
+	newviewblock := createBlockWithQC(es.newview.View, es.newview.View-1)
+	es.forks.blocks[newviewblock.BlockID] = newviewblock
+
+	// I'm the next leader
+	es.committee.leaders[es.newview.View+1] = struct{}{}
+
+	// no QC for the new view
+	err := es.eventhandler.OnQCConstructed(qc)
```

Comment:
I think some tests from the past version collapse to the scenario in Consensus Voting V2: 
* previously, the event handler would process votes. In multiple of the following tests, the setup is similar
   * I have a block for the past view: https://github.com/onflow/flow-go/blob/33372766853dcf6c2e62dcb4aaff52a287285cd7/consensus/hotstuff/eventhandler/event_handler_test.go#L479
   * and we have the ability to produce a QC for the past view if we get another vote: https://github.com/onflow/flow-go/blob/33372766853dcf6c2e62dcb4aaff52a287285cd7/consensus/hotstuff/eventhandler/event_handler_test.go#L481
*  we now trigger a view change by receiving another vote for the past view https://github.com/onflow/flow-go/blob/33372766853dcf6c2e62dcb4aaff52a287285cd7/consensus/hotstuff/eventhandler/event_handler_test.go#L496
   1. we are still missing votes to propose right away (test `TestInNewView_NotLeader_HasBlock_NoVote_IsNextLeader_NoQC`)
   2. we could already have enough votes stored to propose right away (`TestInNewView_NotLeader_HasBlock_NoVote_IsNextLeader_QCBuilt_ViewChanged`)

I think it deserves some thought how to properly translate the tests to the new scenario. I am wondering whether these two tests might collapse into the same test case. 

In my mind, the most critical aspect to cover in this test is:
* Consider Consensus node `Alice` that is currently in view `V`, where `Alice` is leader for view `V +1`
* When Alice is in view `curView = V` and receives the block for `curView` (i.e. block `b1`), Alice remains in view `curView` to collect votes:
   * On the happy path, she gets enough votes and constructs a QC, i.e. she is in view `curView` and  her vote aggregator calls [`OnQCConstructed(qc)`](https://github.com/onflow/flow-go/blob/33372766853dcf6c2e62dcb4aaff52a287285cd7/consensus/hotstuff/eventhandler/event_handler_v2.go#L67) where `qc.View == curView`
   * What should happen is that by receiving the QC, Alice's  PaceMaker  transitions to the new view `curView <- V+1`,  where she will construct her block proposal using the QC just received. 
   * I believe the current implementation already does all this (see [here](https://github.com/onflow/flow-go/blob/33372766853dcf6c2e62dcb4aaff52a287285cd7/consensus/hotstuff/eventhandler/event_handler_v2.go#L407) and [here](https://github.com/onflow/flow-go/blob/33372766853dcf6c2e62dcb4aaff52a287285cd7/consensus/hotstuff/pacemaker/pacemaker.go#L80-L88)). Though, we don't test it. Instead, we test a scenario that cannot occur in practise, because by receiving block `b2`, the event handler would already transition to view `curView+1`

Reply:
>In my mind, the most critical aspect to cover in this test is:


I think the test cases what you described are already covered: 
1. TestOnReceiveProposal_ForCurView_NoVote_IsNextLeader_NoQC
2. TestOnReceiveProposal_ForCurView_NoVote_IsNextLeader_QCBuilt_ViewChange